### PR TITLE
feat: add tree and start on tests for op state ret

### PR DESCRIPTION
**Pull request by gpsanant**
Adds tests for the Operator State Retriever

### CODE REVIEW

File: test/tree/OperatorStateRetrieverUnit.tree [hunk 1]
```diff
@@ -0,0 +1,37 @@
+.
+└── OperatorStateRetriever tree (*** denotes that integration tests are needed to validate path)
+    ├── when getOperatorState(IRegistryCoordinator,bytes32,uint32) is called
+    │   ├── when the given operator has never registered with the given registryCoordinator (their quorumBitmapHistory has length 0)
+    │   │   └── it should revert with "RegistryCoordinator.getQuorumBitmapIndicesAtBlockNumber: operator has no bitmap history at blockNumber"
+    │   └── when the given operator has registered for the first time with the given registryCoordinator
```
Comment (ChaoticWalrus) [inline test/tree/OperatorStateRetrieverUnit.tree:6]: what is the 'for the first time' part of this?
are you just trying to say 'has ever registered' in contrast to the above 'has never registered' ?
I have been writing these trees as assuming that if you make it down a branch, you negate any earlier branching conditions.
Ex:
-- given X
└── yadda yadda yadda (assumes *not* given X)

idk if other people have really been doing this / I appreciate explicitness though!

Comment (ChaoticWalrus) [reply test/tree/OperatorStateRetrieverUnit.tree:6]: ahhhh nevermind, I see what you're going for here. it's either
(a) 'the given operator registered for the first time with the given registryCoordinator after the given blockNumber'
OR
(b) 'the given operator registered for the first time with the given registryCoordinator *at or* before the given blockNumber'

File: test/tree/OperatorStateRetrieverUnit.tree [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    │       │   └── it should revert with "RegistryCoordinator.getQuorumBitmapIndicesAtBlockNumber: operator has no bitmap history at blockNumber"
+    │       └── before the given blockNumber
+    │           ├── and was not registered for any quorums at the given blockNumber
+    │           │   └── it should return (0, [])
+    │           └── and was registered for a set of quorums at the given blockNumber
+    │               └── it should return their quorum bitmap as the first argument and a list of the list of operators (by id and stake) ordered by index for each quorum in the set ascending by quorumNumber
+    ├── when getOperatorState(IRegistryCoordinator,bytes memory,uint32) is called
+    │   ├── when at any of the given quorumNumbers are not created at the time of call
+    │   │   └── it should revert with "IndexRegistry._operatorCountAtBlockNumber: quorum does not exist"
+    │   ├── when at any of the given quorumNumbers are not created before or at the given blockNumber
```
Comment (ChaoticWalrus) [inline test/tree/OperatorStateRetrieverUnit.tree]: 'are'->'were' I think?

File: test/tree/OperatorStateRetrieverUnit.tree [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    │   │   └── it should revert with "IndexRegistry._operatorCountAtBlockNumber: quorum did not exist at given block number"
+    │   └── when the given quorumNumbers are created on the given registry registryCoordinator
+    │       └── it should return their quorum bitmap as the first argument and a list of the list of operators (by id and stake) ordered by index for each quorum in the set ascending by quorumNumber
```
Comment (ChaoticWalrus) [inline test/tree/OperatorStateRetrieverUnit.tree:22]: the 'their' here is super ambiguous. I think you mean `the given operator's`?

Comment (ChaoticWalrus) [reply test/tree/OperatorStateRetrieverUnit.tree:22]: also maybe `_current_ quorum bitmap`?

File: test/unit/OperatorStateRetrieverUnit.t.sol [hunk 4]
```diff
@@ -6,8 +6,238 @@ import "../utils/MockAVSDeployer.sol";
...
+        assertEq(checkSignaturesIndices.quorumApkIndices[1], 2);
+        assertEq(checkSignaturesIndices.totalStakeIndices.length, 2);
+        assertEq(checkSignaturesIndices.totalStakeIndices[0], 3);
+        assertEq(checkSignaturesIndices.totalStakeIndices[1], 2);
+        assertEq(checkSignaturesIndices.nonSignerStakeIndices.length, 2);
+        assertEq(checkSignaturesIndices.nonSignerStakeIndices[0].length, 1);
+        assertEq(checkSignaturesIndices.nonSignerStakeIndices[0][0], 0);
+        assertEq(checkSignaturesIndices.nonSignerStakeIndices[1].length, 2);
+        assertEq(checkSignaturesIndices.nonSignerStakeIndices[1][0], 0);
+        assertEq(checkSignaturesIndices.nonSignerStakeIndices[1][1], 0);
```
Comment (ChaoticWalrus) [inline test/unit/OperatorStateRetrieverUnit.t.sol]: ngl it's quite hard to tell if this is all correct.
I appreciate how much is being checked but...can you maybe add some comments like:
```
// returned nonSignerQuorumBitmapIndices should be [2, 0]
// quorumApkIndices should be...
```
or perhaps what I'm missing is a little more context above?
a couple comments could be helpful, like:
```
// register defaultOperator for the 0-indexed quorum.
// expect a new operator bitmap entry (length now 1),
// new quorumApk entry (length now 1),
// new operator stake entry (length now 1),
// and new total stake entry (length now 2)
```

**Review by ChaoticWalrus — CHANGES_REQUESTED** [review]
Changes look good, I'm just struggling to understand one test. Can you add more comments or details to make it understandable? There is too much to keep straight in my brain here.

File: test/unit/OperatorStateRetrieverUnit.t.sol [hunk 5]
```diff
@@ -6,8 +6,238 @@ import "../utils/MockAVSDeployer.sol";
...
+        bytes32 otherOperatorId = BN254.hashG1Point(otherPubKey);
+        _registerOperatorWithCoordinator(otherOperator, quorumBitmapThree, otherPubKey, defaultStake -1);
+
+        OperatorStateRetriever.Operator[][] memory operators = operatorStateRetriever.getOperatorState(registryCoordinator, BitmapUtils.bitmapToBytesArray(quorumBitmapThree), uint32(block.number));
+        assertEq(operators.length, 2, "operators are registered for 2 quorums, so there should be 2 arrays of operators");
+        assertEq(operators[0].length, 2, "operators are registered for 2 quorums, so there should be 1 operator in the first quorum");
+        assertEq(operators[1].length, 1, "operators are registered for 2 quorums, so there should be 1 operator in the second quorum");
+        assertEq(operators[0][0].operatorId, defaultOperatorId, "the first operator in the first quorum should be the default operator");
+        assertEq(operators[0][0].stake, defaultStake, "the first operator in the first quorum should have the default stake");
+        assertEq(operators[0][1].operatorId, otherOperatorId, "the second operator in the first quorum should be the other operator");
```
Comment (8sunyuan) [inline test/unit/OperatorStateRetrieverUnit.t.sol]: For the `Operator` state checks, could you pull/rebase from m2-mainnet and add operator address checks as well?
I made a small addition to this contract for avs-sync here https://github.com/Layr-Labs/eigenlayer-middleware/pull/143

File: test/tree/OperatorStateRetrieverUnit.tree (hunk 2 above)
Comment (gpsanant) [reply test/tree/OperatorStateRetrieverUnit.tree]: [3a1b807](https://github.com/Layr-Labs/eigenlayer-middleware/pull/140/commits/3a1b80758dc15258727f0e527415ea94200d00b5)

File: test/tree/OperatorStateRetrieverUnit.tree (hunk 3 above)
Comment (gpsanant) [reply test/tree/OperatorStateRetrieverUnit.tree:22]: [3a1b807](https://github.com/Layr-Labs/eigenlayer-middleware/pull/140/commits/3a1b80758dc15258727f0e527415ea94200d00b5)

File: test/unit/OperatorStateRetrieverUnit.t.sol (hunk 4 above)
Comment (gpsanant) [reply test/unit/OperatorStateRetrieverUnit.t.sol]: here's some comments: [66fffb0](https://github.com/Layr-Labs/eigenlayer-middleware/pull/140/commits/66fffb070709fe57f8f03c508a8d161c80b8ef49)

File: test/unit/OperatorStateRetrieverUnit.t.sol (hunk 5 above)
Comment (gpsanant) [reply test/unit/OperatorStateRetrieverUnit.t.sol]: [0d48200](https://github.com/Layr-Labs/eigenlayer-middleware/pull/140/commits/0d48200ca01817b91fa3d062838382973c63ed1b)

Comment (gpsanant) [reply test/unit/OperatorStateRetrieverUnit.t.sol]: done! have you checked that getOperatorState still runs within a 50million gas with this change and large operator sets?

**Review by ChaoticWalrus — APPROVED** [review]

**Comment by ChaoticWalrus** [comment]
New comments are 🐐 
ty ty

**Comment by ChaoticWalrus** [comment]
I noted in DMs but some of the links to specific commits that you've posted above seem to be broken.
Perhaps something having to do with the force-push action you took?

**Pull request merged**